pkg/server/commands/reverseexpand/reverse_expand.go

```diff
@@ -167,11 +167,16 @@ type ReverseExpandResult struct {
 
 type ResolutionMetadata struct {
 	QueryCount *uint32
+
+	// The number of times we are expanding from each node to find set of objects
+	// In case of intersection and exclusion, it also includes the dispatch counts done for Check calls to exclude Objects
+	DispatchCount *uint32
```

Comment:
```suggestion
	// The number of times we are expanding from each node to find set of objects
	DispatchCount *uint32
```

ReverseExpand is not aware that any Checks are happening. That's ListObjects's responsibility.

We should introduce a new `ResolutionMetadata` specifically for ListObjects, so we can distinguish between the two:

```go
type ListObjectsResponse struct {
	Objects            []string
	ResolutionMetadata ListObjectsResolutionMetadata
}

type ListObjectsResolutionMetadata struct {
	QueryCount *uint32

	// The number of times we are expanding from each node to find set of objects, 
       // AND the dispatch counts done for Check calls to exclude Objects
	DispatchCount *uint32
}

func NewResolutionMetadata() *ListObjectsResolutionMetadata {
	return &ListObjectsResolutionMetadata{
		QueryCount:    new(uint32),
		DispatchCount: new(uint32),
	}
}
```

you will need to make some changes, for example:

1. `evaluate`'s last parameter needs to change its type to `ListObjectsResolutionMetadata`

https://github.com/openfga/openfga/blob/b351a720134d323a5d64b9e62cffca9dd95b3e94/pkg/server/commands/list_objects.go#L165-L171

2. here, at the end of the goroutine, https://github.com/openfga/openfga/blob/b351a720134d323a5d64b9e62cffca9dd95b3e94/pkg/server/commands/list_objects.go#L255-L269 you'll need to add `atomic.AddUint32(listObjectsResolutionMetadata.DispatchCount, *reverseexpandResolutionMetadata.DispatchCount)`

Reply:
Agreed, but is there a future use case where reverse expand will be reused for another operation? In that case it is better to leave it here to be reused. 

Or else we should move it to ListObjects

Reply:
Also looks like Check does the same and has it inside graph.go - https://github.com/openfga/openfga/blob/main/internal/graph/graph.go#L44

Reply:
Maria can you explain why you want to separate? Is it just because of the code comment?

Reply:
> Also looks like Check does the same and has it inside graph.go - [main/internal/graph/graph.go#L44](https://github.com/openfga/openfga/blob/main/internal/graph/graph.go?rgh-link-date=2024-03-08T11%3A11%3A34Z#L44)

Yeah that is okay, the resolution metadata of Check only includes Check resolutions

> Maria can you explain why you want to separate? Is it just because of the code comment?

Single responsibility principle. I think the DispatchCount of ReverseExpand should only include the number of dispatches, not the Checks that happen outside of it

Reply:
As for the godoc comment, can we just simply say 

> The number of subproblems that had to be dispatched to resolve the parent subproblem.

We're defining `Dispatch` in terms of nodes and other vernacular. It's confusing at best.

Reply:
> Single responsibility principle. I think the DispatchCount of ReverseExpand should only include the number of dispatches, not the Checks that happen outside of it

I agree with this. Guess I misunderstood to use the same ResolutionMetadata. It would be better to use 2. 1 for reverse_expand and another for ListObjects 👍 I'll make those changes

Reply:
@miparnisari made the changes you suggested. Check it out when you have sometime